testsuite/testcases/src/validator_join_leave_test.rs

```diff
@@ -0,0 +1,257 @@
+                .await
+        })?;
+
+        // Stop and reset 1/3 validators
+        info!("Make 1/3 validators leave the validator set!");
+        for operator_index in validator_cli_indices.iter().take(num_validators / 3) {
+            runtime.block_on(async {
+                cli.leave_validator_set(*operator_index, None)
+                    .await
+                    .unwrap();
+
+                reconfig(
+                    &rest_client,
+                    &transaction_factory,
+                    ctx.swarm().chain_info().root_account(),
+                )
+                .await
+            });
+        }
+
+        // Restart the validators.
+        for operator_index in validator_cli_indices.iter().take(num_validators / 3) {
+            runtime.block_on(async {
+                cli.join_validator_set(*operator_index, None).await.unwrap();
+
+                reconfig(
+                    &rest_client,
+                    &transaction_factory,
+                    ctx.swarm().chain_info().root_account(),
+                )
+                .await
+            });
+        }
+
+        // Wait for all nodes to synchronize and stabilize.
+        info!("Waiting for the validators to be synchronized.");
+        runtime.block_on(async {
+            ctx.swarm()
+                .wait_for_all_nodes_to_catchup(Duration::from_secs(MAX_NODE_LAG_SECS))
+                .await
+        })?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum ValidatorState {
+    ACTIVE,
+    JOINING,
+    LEAVING,
+    NONE,
+}
+
+struct ValidatorNodeKeys {
+    account_private_key: Ed25519PrivateKey,
+    network_private_key: x25519::PrivateKey,
+    consensus_private_key: bls12381::PrivateKey,
+}
+
+impl ValidatorNodeKeys {
+    pub fn new(keygen: &mut KeyGen) -> Self {
+        Self {
+            account_private_key: keygen.generate_ed25519_private_key(),
+            network_private_key: keygen.generate_x25519_private_key().unwrap(),
+            consensus_private_key: keygen.generate_bls12381_private_key(),
+        }
+    }
+
+    pub fn network_public_key(&self) -> x25519::PublicKey {
+        self.network_private_key.public_key()
+    }
+
+    pub fn consensus_public_key(&self) -> bls12381::PublicKey {
+        bls12381::PublicKey::from(&self.consensus_private_key)
+    }
+
+    pub fn consensus_proof_of_possession(&self) -> bls12381::ProofOfPossession {
+        bls12381::ProofOfPossession::create(&self.consensus_private_key)
+    }
+}
+
+async fn init_validator_account(
+    cli: &mut CliTestFramework,
+    keygen: &mut KeyGen,
+    amount: Option<u64>,
+) -> (usize, ValidatorNodeKeys) {
+    let validator_node_keys = ValidatorNodeKeys::new(keygen);
+    let validator_cli_index = cli
+        .create_cli_account_from_faucet(validator_node_keys.account_private_key.clone(), amount)
+        .await
+        .unwrap();
+
+    cli.assert_account_balance_now(validator_cli_index, amount.unwrap_or(DEFAULT_FUNDED_COINS))
```

Comment:
this will be default value from genesis, but you don't need to assert this really.

you might need to actually see if genesis gives them any extra coins, for gas fees for transactions you will be issuing